### PR TITLE
EP documentation updates

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Add some initial documentation for the CUDA EP with some details about configuration options.
Update Nuphar EP documentation usage of deprecated set_nuphar_settings() function.

**Motivation and Context**
Needed somewhere to document CUDA EP provider-specific options.
Deprecation of Python global configuration functions.

### CODE REVIEW

File: docs/reference/execution-providers/CUDA-ExecutionProvider.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+---
+title: CUDA
+parent: Execution Providers
+grand_parent: Reference
+nav_order: 1
+---
+
+# CUDA Execution Provider
+
+The CUDA Execution Provider enables hardware accelerated computation on CUDA-enabled GPUs.
```
Comment (jywu-msft) [inline docs/reference/execution-providers/CUDA-ExecutionProvider.md:10]: "computation on Nvidia GPUs" might be better?

File: docs/reference/execution-providers/CUDA-ExecutionProvider.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+## Build
+For build instructions, please see the [BUILD page](../../how-to/build.md#CUDA).
+
+## Configuration Options
+The CUDA EP supports the following configuration options:
```
Comment (jywu-msft) [inline docs/reference/execution-providers/CUDA-ExecutionProvider.md:16]: would it be possible to link to some sample usage in C++/python etc. ?

File: docs/reference/execution-providers/CUDA-ExecutionProvider.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+### device_id
+The device ID.
+
+### cuda_mem_limit
+The memory limit of the device memory arena in bytes.
```
Comment (pranavsharma) [inline docs/reference/execution-providers/CUDA-ExecutionProvider.md:22]: This might need more clarification based on this question https://github.com/microsoft/onnxruntime/issues/6215.

**Review by jywu-msft — APPROVED** [review]

**Review by KeDengMS — APPROVED** [review]
:shipit:

File: docs/reference/execution-providers/CUDA-ExecutionProvider.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
+---
+title: CUDA
```
Comment (hariharans29) [inline docs/reference/execution-providers/CUDA-ExecutionProvider.md:2]: Should this file also live in master ? under docs/execution_providers ? It seems like the other files are present there as well ?

Comment (edgchen1) [reply docs/reference/execution-providers/CUDA-ExecutionProvider.md:2]: this PR is removing them from master: https://github.com/microsoft/onnxruntime/pull/6225
i think this is the official location.

File: docs/reference/execution-providers/CUDA-ExecutionProvider.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import onnxruntime as ort
+
+model_path = '<path to model>'
+
+providers = [
+    ('CUDAExecutionProvider', {
+        'device_id': 0,
+        'arena_extend_strategy': 'kNextPowerOfTwo',
+        'cuda_mem_limit': 2 * 1024 * 1024 * 1024,
+        'cudnn_conv_algo_search': 'EXHAUSTIVE',
```
Comment (hariharans29) [inline docs/reference/execution-providers/CUDA-ExecutionProvider.md:58]: I think the key you want in the example is `cudnn_conv_algo`. reading the code, that is the key it is going to be looking for

Comment (hariharans29) [inline docs/reference/execution-providers/CUDA-ExecutionProvider.md:58]: Btw, I am making some changes that needs to internally use CUDAExecutionProviderInfo::FromProviderOptions(), when I change the key to 'cudnn_conv_algo' (whcih is the right key) and pass in any legitimate values: EXHAUSTIVE, HEURISTIC, DEFAULT, the parsing method simply crashes. Is the method tested somewhere ? I suspect when you correct the key above and try the python API to configure the CUDA EP, it will crash as well.

`arena_extend_strategy` works seems to work fine for all valid values.

File: docs/reference/execution-providers/CUDA-ExecutionProvider.md (hunk 4 above)
Comment (hariharans29) [reply docs/reference/execution-providers/CUDA-ExecutionProvider.md:2]: Thanks, got it.

File: docs/reference/execution-providers/CUDA-ExecutionProvider.md (hunk 5 above)
Comment (edgchen1) [reply docs/reference/execution-providers/CUDA-ExecutionProvider.md:58]: it was just updated by this PR: [https://github.com/microsoft/onnxruntime/pull/6171](https://github.com/microsoft/onnxruntime/pull/6171)

---
In reply to: [553438408](https://github.com/microsoft/onnxruntime/pull/6253#discussion_r553438408) [](ancestors = 553438408)

Comment (edgchen1) [reply docs/reference/execution-providers/CUDA-ExecutionProvider.md:58]: please update and try again, it should work now.
added some tests recently: [https://github.com/microsoft/onnxruntime/blob/481a2cdf614458b550612176655656c4ec69b112/onnxruntime/test/python/onnxruntime_test_python.py#L120](https://github.com/microsoft/onnxruntime/blob/481a2cdf614458b550612176655656c4ec69b112/onnxruntime/test/python/onnxruntime_test_python.py#L120)

---
In reply to: [553452048](https://github.com/microsoft/onnxruntime/pull/6253#discussion_r553452048) [](ancestors = 553452048)

Comment (hariharans29) [reply docs/reference/execution-providers/CUDA-ExecutionProvider.md:58]: Ok

Comment (hariharans29) [reply docs/reference/execution-providers/CUDA-ExecutionProvider.md:58]: After pulling in the latest changes, all valid values seem to work fine, thanks!

File: docs/reference/execution-providers/CUDA-ExecutionProvider.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+For build instructions, please see the [BUILD page](../../how-to/build.md#CUDA).
+
+## Configuration Options
+The CUDA Execution Provider supports the following configuration options.
+
+### device_id
+The device ID.
+
+### cuda_mem_limit
+The size limit of the device memory arena in bytes. This size limit is only for the execution provider's arena. The total device memory usage may be higher.
```
Comment (hariharans29) [inline docs/reference/execution-providers/CUDA-ExecutionProvider.md:22]: Sorry for adding another comment. Maybe it is worth specifying the "default" values for each of these parameters ? For example, reading the doc, I understand what the valid values for `cudnn_conv_algo_search` are and what each option means. What is hard to parse is the "default" value for each.

**Review by hariharans29 — APPROVED** [review]

**Review by natke — APPROVED** [review]
